jax/experimental/mosaic/gpu/fragmented_array.py

```diff
@@ -315,30 +316,66 @@ def max(self, other):
   def exp(self, approx: bool = False):
     if not ir.FloatType.isinstance(self.mlir_dtype):
       raise NotImplementedError
-    def fast_exp(x):
```

Comment:
From 6bc7929376161fe77f4f5e30b44e7fb752c12a1b Mon Sep 17 00:00:00 2001
From: Adam Paszke <apaszke@google.com>
Date: Wed, 24 Jul 2024 01:14:13 -0700
Subject: [PATCH] [Mosaic GPU] Add sin/cos + unify support for approximate
 transcendental functions

PiperOrigin-RevId: 655469213
---
 .../mosaic/gpu/fragmented_array.py            | 57 +++++++++++++++----
 tests/mosaic/gpu_test.py                      | 17 +++++-
 2 files changed, 61 insertions(+), 13 deletions(-)

diff --git a/jax/experimental/mosaic/gpu/fragmented_array.py b/jax/experimental/mosaic/gpu/fragmented_array.py
index 1decd8acacf2..8b13a00bced9 100644
--- a/jax/experimental/mosaic/gpu/fragmented_array.py
+++ b/jax/experimental/mosaic/gpu/fragmented_array.py
@@ -15,6 +15,7 @@
 """Utilities for code generator."""
 
 import dataclasses
+from typing import Callable
 
 import jax
 from jaxlib.mlir import ir
@@ -315,30 +316,66 @@ def max(self, other):
   def exp(self, approx: bool = False):
     if not ir.FloatType.isinstance(self.mlir_dtype):
       raise NotImplementedError
-    def fast_exp(x):
+    if approx:
       f32 = ir.F32Type.get()
       if self.mlir_dtype != f32:
         raise NotImplementedError
       log2e = arith.constant(f32, ir.FloatAttr.get(f32, 1.4426950408889634))
-      if x.type == f32:
+      def fast_exp(x):
         scaled = arith.mulf(x, log2e)
-        return llvm.inline_asm(
-            f32, [scaled], "ex2.approx.f32 $0,$1;", "=f,f", asm_dialect=0
-        )
+        return llvm.inline_asm(f32, [scaled], "ex2.approx.f32 $0, $1;", "=f,f")
+      return self._pointwise(self._lift_fast_unary(fast_exp))
+    return self._pointwise(mlir_math.exp)
+
+  def sin(self, approx: bool = False):
+    if not ir.FloatType.isinstance(self.mlir_dtype):
+      raise NotImplementedError
+    if approx and self.mlir_dtype != ir.F32Type.get():
+      raise NotImplementedError
+    return self._pointwise(
+        self._lift_fast_unary("sin.approx.f32") if approx else mlir_math.sin
+    )
+
+  def cos(self, approx: bool = False):
+    if not ir.FloatType.isinstance(self.mlir_dtype):
+      raise NotImplementedError
+    if approx and self.mlir_dtype != ir.F32Type.get():
+      raise NotImplementedError
+    return self._pointwise(
+        self._lift_fast_unary("cos.approx.f32") if approx else mlir_math.cos
+    )
+
+  def rsqrt(self, approx: bool = False):
+    if not ir.FloatType.isinstance(self.mlir_dtype):
+      raise NotImplementedError
+    if approx and self.mlir_dtype != ir.F32Type.get():
+      raise NotImplementedError
+    return self._pointwise(
+        self._lift_fast_unary("rsqrt.approx.f32") if approx else mlir_math.rsqrt
+    )
+
+  @staticmethod
+  def _lift_fast_unary(
+      instr: str | Callable[[ir.Value], ir.Value],
+  ) -> Callable[[ir.Value], ir.Value]:
+    def fast_instr(x):
+      f32 = ir.F32Type.get()
+      if x.type == f32:
+        if isinstance(instr, str):
+          return llvm.inline_asm(f32, [x], instr + " $0, $1;", "=f,f")
+        else:
+          return instr(x)
       elif ir.VectorType.isinstance(x.type):
         index = ir.IndexType.get()
         result = llvm.mlir_undef(x.type)
         for i in range(2):
           v = vector.extractelement(x, position=c(i, index))
-          vr = fast_exp(v)
+          vr = fast_instr(v)
           result = vector.insertelement(vr, result, position=c(i, index))
         return result
       else:
         raise NotImplementedError(x.type)
-    return self._pointwise(fast_exp if approx else mlir_math.exp)
-
-  def rsqrt(self):
-    return self._pointwise(mlir_math.rsqrt)
+    return fast_instr
 
   def __and__(self, other):
     if not ir.IntegerType.isinstance(self.mlir_dtype):
diff --git a/tests/mosaic/gpu_test.py b/tests/mosaic/gpu_test.py
index 45c948abd00b..fe66c6fa49bd 100644
--- a/tests/mosaic/gpu_test.py
+++ b/tests/mosaic/gpu_test.py
@@ -987,12 +987,21 @@ def kernel(ctx, dst, _):
         np.testing.assert_array_equal(result, np_op(ref_x, ref_rhs))
 
   @parameterized.product(
-      ops=((lambda x: mgpu.FragmentedArray.exp(x), np.exp),),
+      ops=(
+          (lambda x: mgpu.FragmentedArray.exp(x), np.exp, False),
+          (lambda x: mgpu.FragmentedArray.exp(x, approx=True), np.exp, True),
+          (lambda x: mgpu.FragmentedArray.sin(x), np.sin, False),
+          (lambda x: mgpu.FragmentedArray.sin(x, approx=True), np.sin, True),
+          (lambda x: mgpu.FragmentedArray.cos(x), np.cos, False),
+          (lambda x: mgpu.FragmentedArray.cos(x, approx=True), np.cos, True),
+          (lambda x: mgpu.FragmentedArray.rsqrt(x), jax.lax.rsqrt, False),
+          (lambda x: mgpu.FragmentedArray.rsqrt(x, approx=True), jax.lax.rsqrt, True),
+      ),
       m=(64, 128),
       n=(8, 16, 32, 64, 80, 128, 256),
   )
   def test_unary(self, ops, m=64, n=32):
-    op, np_op = ops
+    op, np_op, is_approx = ops
     def kernel(ctx, dst, _):
       f32 = ir.F32Type.get()
       iota = iota_tensor(m=m, n=n, mlir_dtype=f32)
@@ -1002,7 +1011,9 @@ def kernel(ctx, dst, _):
         kernel, (1, 1, 1), (128, 1, 1), (), out_shape, ()
     )()
     x = np.arange(m * n, dtype=jnp.float32).reshape(m, n)
-    np.testing.assert_allclose(result, np_op(x), atol=2e-7, rtol=2e-7)
+    atol = 5e-3 if is_approx else 2e-7
+    rtol = 4e-6 if is_approx else 2e-7
+    np.testing.assert_allclose(result, np_op(x), atol=atol, rtol=rtol)
 
   @parameterized.product(
       op=(arith.addf, arith.maximumf),